### PR TITLE
migrate more rest handlers to select by resource enablement

### PR DESCRIPTION
**Pull request by deads2k**
In support of, "no new beta APIs enabled by default"

/kind clean-up
/sig api-machinery
/priority important-soon
/assign @liggitt

```release-note
NONE
```

### CODE REVIEW

File: pkg/controlplane/instance.go [hunk 1]
```diff
@@ -392,24 +392,9 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 	}
 
 	// install legacy rest storage
-	if c.ExtraConfig.APIResourceConfigSource.VersionEnabled(apiv1.SchemeGroupVersion) {
-		legacyRESTStorageProvider := corerest.LegacyRESTStorageProvider{
```
Comment (deads2k) [inline pkg/controlplane/instance.go:396]: moved into InstallLegacyAPI to make it impossible for someone to rely on it later in this method.

File: pkg/registry/core/rest/storage_core.go [hunk 2]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		if utilfeature.DefaultFeatureGate.Enabled(features.EphemeralContainers) {
+			storage[resource+"/ephemeralcontainers"] = podStorage.EphemeralContainers
+		}
+
+	}
+	if resource := "bindings"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
+		storage[resource] = podStorage.LegacyBinding
+	}
+
+	if resource := "podTemplates"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:299]: These have to change to lowercase if we're comparing against enablement by `--runtime-config`. We normalize this to lower-case in `GenericAPIServer#getAPIGroupVersion` and `DefaultStorageFactory#ResourcePrefix`, and forbid upper-case characters in `--runtime-config` in `MergeAPIResourceConfigs`

```suggestion
	if resource := "podtemplates"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

I would maybe change the normalization to start erroring in `GenericAPIServer#getAPIGroupVersion` to flag bad setup

File: pkg/registry/core/rest/storage_core.go [hunk 3]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = podTemplateStorage
+	}
+
+	if resource := "replicationControllers"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:303]:
```suggestion
	if resource := "replicationcontrollers"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 4]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = controllerStorage.Controller
+		storage[resource+"/status"] = controllerStorage.Status
+		if legacyscheme.Scheme.IsVersionRegistered(schema.GroupVersion{Group: "autoscaling", Version: "v1"}) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:306]: pre-existing, I know, but checking scheme registration here is so weird...

File: pkg/registry/core/rest/storage_core.go [hunk 5]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
-		"replicationControllers":        controllerStorage.Controller,
-		"replicationControllers/status": controllerStorage.Status,
+	if resource := "events"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
+		storage[resource] = eventStorage
+	}
 
-		"services":        serviceRESTStorage,
-		"services/proxy":  serviceRESTProxy,
-		"services/status": serviceStatusStorage,
+	if resource := "limitRanges"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:331]:
```suggestion
	if resource := "limitranges"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 6]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = limitRangeStorage
+	}
 
-		"endpoints": endpointsStorage,
+	if resource := "resourceQuotas"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:335]:
```suggestion
	if resource := "resourcequotas"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 7]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
-		"namespaces/status":             namespaceStatusStorage,
-		"namespaces/finalize":           namespaceFinalizeStorage,
-		"secrets":                       secretStorage,
-		"serviceAccounts":               serviceAccountStorage,
-		"persistentVolumes":             persistentVolumeStorage,
-		"persistentVolumes/status":      persistentVolumeStatusStorage,
-		"persistentVolumeClaims":        persistentVolumeClaimStorage,
-		"persistentVolumeClaims/status": persistentVolumeClaimStatusStorage,
-		"configMaps":                    configMapStorage,
+	if resource := "serviceAccounts"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:350]:
```suggestion
	if resource := "serviceaccounts"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 8]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = serviceAccountStorage
+		if serviceAccountStorage.Token != nil {
+			storage[resource+"/token"] = serviceAccountStorage.Token
+		}
+	}
 
-		"componentStatuses": componentstatus.NewStorage(componentStatusStorage{c.StorageFactory}.serversToValidate),
+	if resource := "persistentVolumes"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:357]:
```suggestion
	if resource := "persistentvolumes"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 9]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = persistentVolumeStorage
+		storage[resource+"/status"] = persistentVolumeStatusStorage
 	}
-	if legacyscheme.Scheme.IsVersionRegistered(schema.GroupVersion{Group: "autoscaling", Version: "v1"}) {
-		restStorageMap["replicationControllers/scale"] = controllerStorage.Scale
+
+	if resource := "persistentVolumeClaims"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:362]:
```suggestion
	if resource := "persistentvolumeclaims"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 10]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = persistentVolumeClaimStorage
+		storage[resource+"/status"] = persistentVolumeClaimStatusStorage
 	}
-	if podStorage.Eviction != nil {
-		restStorageMap["pods/eviction"] = podStorage.Eviction
+
+	if resource := "configMaps"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:367]:
```suggestion
	if resource := "configmaps"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go [hunk 11]
```diff
@@ -273,63 +274,107 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
...
+		storage[resource] = configMapStorage
 	}
-	if serviceAccountStorage.Token != nil {
-		restStorageMap["serviceaccounts/token"] = serviceAccountStorage.Token
+
+	if resource := "componentStatuses"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```
Comment (liggitt) [inline pkg/registry/core/rest/storage_core.go:371]:
```suggestion
	if resource := "componentstatuses"; apiResourceConfigSource.ResourceEnabled(corev1.SchemeGroupVersion.WithResource(resource)) {
```

File: pkg/registry/core/rest/storage_core.go (hunk 4 above)
Comment (liggitt) [reply pkg/registry/core/rest/storage_core.go:306]: I would drop the scheme registration check. if anything, I would check if the Scale field is non-nil

**Comment by deads2k** [comment]
squashed.

/hold cancel

**Pull request merged**